clrtrust

```diff
@@ -91,6 +102,31 @@ find_all_certs() {
     find_certs $CLR_LOCAL_TRUST_SRC/trusted
 }
 
+# a c_rehash implementation for creating openssl-style CApath. this
+# implementation is much simpler than openssl's one: it is designed to rehash a
+# newly created store. it only takes directory to process as the argument and
+# no options. it further assumes that it runs on a pristine directory where
+# every file is a valid certificate, no duplicates (clrtrust ensures it prior to
+# calling this function).
+c_rehash_internal() {
+    local dir=$1
+    if [ -z "$dir" ] || [ ! -d "$dir" ] || [ ! -w "$dir" ]; then
+        return 1
+    fi
+    (
+        cd "$dir"
+        find . -maxdepth 1 -type f | while read f; do
+            name=$(openssl x509 -in "$f" -noout -subject_hash 2>/dev/null)
+            lnno=0
+            while [ -h "${name}.${lnno}" ]; do
```

Comment:
Obviously there is a bit of a race here. If you can live with it then fine, otherwise I suggest

```
 lnno=0
 while ! ln -s "$f" "${name}.${lnno}"
 do ((lnno++))
 done 2>/dev/null
```

and use the fact that 'ln` fails if it tries to make a link.

For an extra bonus point, unset VERSION_CONTROL in the script to stop ln from making backups.

Reply:
This is always done on a staging area, so each instance of `clrtrust` will get its own directory (via `mktemp -d`) to rehash before its swapped with the store. I tried to outline some assumption this impl makes at the beginning of the function, but failed to mention this "isolation" part.

Still, I will adopt your suggestion, it's elegant. And will add a few words on "isolation" assumption.

Reply:
If it is isolated, then just ignore this, As I said probably going to get to the heat death of the universe before this happens in practice. Still it is worth while knowing about the technique.

Reply:
Thought prompted by your comment.

If this happens in isolation, before being "swapped", do you take a lock of some kind to stop 2 things doing their own changes in parallel (but isolated) and then one "commits" the change and then the other **replaces** the first change with their own change?

(I omitted the error handling for what happens if 100 attempts to make a symlink fail of course)

Reply:
@icarus-sparry, no, locking is not yet done (even though i plan to). in fairness, i did my best to minimize the writing of the store section, so it's unlikely two concurrent commands would hit race condition. that said, it does not make serializing that "swap" section unnecessary, it will be done. i was thinking `flock` will a proper mechanism and Clear ships it with the bare OS. i filed #9 since i knew the issue was there, but it escaped me. now it won't.

and 100 collisions is a good limit, we do not need to handle more than that.

Reply:
Glad to be of help, rather than just a nuisance.
My point about the 100 is that if something goes wrong, e.g. someone changes the permission on the directory to remove write access so ln can never succeed you want to exit with an error message rather than be spawning an essentially unlimited number of ln commands at high speed.

`flock` is fine, or you can create your own using something like this
```
LOCK=/run/clrtrust.lock
lock(){
   local mylock=${LOCK}.$$ # could use mktemp
   while ! ln "$mylock" "$LOCK"
   do sleep 1
   done
   rm -f "$mylock"
}

unlock() {
   rm -f "$LOCK"
}
```